DOC: add deprecation message to index_update and friends

jakevdp opened the pull request:
Fixes #8064

Preview the new docs here: https://jax--8065.org.readthedocs.build/en/8065/_autosummary/jax.ops.index_update.html#jax.ops.index_update

hawkinsp commented:
This looks good.

There is one further step we probably should do. The `index_mul` etc docs have an additional and necessary paragraph:

"
Unlike the NumPy code x[idx] *= y, if multiple indices refer to the same location the updates will be multiplied. (NumPy would only apply the last update, rather than multiplying the updates.) The order in which conflicting updates are applied is implementation-defined and may be nondeterministic (e.g., due to concurrency on some hardware platforms).
"

We should probaby add that to the `.at[...]` documentation, suitably modified.

hawkinsp commented:
> This looks good.
> 
> There is one further step we probably should do. The `index_mul` etc docs have an additional and necessary paragraph:
> 
> " Unlike the NumPy code x[idx] *= y, if multiple indices refer to the same location the updates will be multiplied. (NumPy would only apply the last update, rather than multiplying the updates.) The order in which conflicting updates are applied is implementation-defined and may be nondeterministic (e.g., due to concurrency on some hardware platforms). "
> 
> We should probaby add that to the `.at[...]` documentation, suitably modified.

There is also some useful parameter documentation on those functions that needs to be transferred. Maybe for a future PR, though.

jakevdp commented:
Added the extra paragraph, as well as a TODO to document additional keyword arguments.

hawkinsp commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -6690,22 +6690,43 @@ class _IndexUpdateHelper:
+
+  Unlike NumPy in-place operations such as :code:`x[idx] += y`, if multiple
+  indices refer to the same location, all updates will be applied (NumPy would
+  only apply the last update, rather than multiplying the updates.) The order
```
"multiplying" only makes sense for `index_mul` perhaps. But instead "applying" might work.

jakevdp replied:
Done.

jakevdp committed:
DOC: move index update API docs to jnp.ndarray.at
- Add docstring to abstract  property
- Add explicit HTML documentation of this property
- Mark index update functions as deprecated, linking to this documentation
